chain/beacon/sync.go

```diff
@@ -3,133 +3,184 @@ package beacon
 import (
 	"context"
 	"errors"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/drand/drand/chain"
 	"github.com/drand/drand/log"
 	"github.com/drand/drand/net"
+	"github.com/drand/drand/protobuf/drand"
 	proto "github.com/drand/drand/protobuf/drand"
 )
 
-// SyncChain is the server side call that reply with the beacon in order to the
-// client requesting the syncing.
-func (h *Handler) SyncChain(req *proto.SyncRequest, p proto.Protocol_SyncChainServer) error {
-	fromRound := req.GetFromRound()
-	addr := net.RemoteAddress(p.Context())
-	last, _ := h.chain.Last()
-	h.l.Debug("received", "sync_request", "from", addr, "from_round", fromRound, "head_at", last.Round)
-	if last.Round < fromRound {
-		return errors.New("no beacon stored above requested round")
+// Syncer allows to follow a chain from other nodes and replies to syncing
+// requests.
+type Syncer interface {
+	// Follow is a blocking call that continuously fetches the beacon from the
+	// given nodes, verify the validity (chain etc) and then  stores it, until
+	// the context is cancelled or the round reaches upTo.  To follow
+	// indefinitely, simply pass upTo = 0.
+	Follow(c context.Context, upTo uint64, to []net.Peer) error
+	// Syncing returns true if the syncer is currently being syncing
+	Syncing() bool
+	// SyncChain imeplements the server side of the syncing process
+	SyncChain(req *proto.SyncRequest, p proto.Protocol_SyncChainServer) error
+}
+
+// syncer implements the Syncer interface
+type syncer struct {
+	l         log.Logger
+	store     CallbackStore
+	info      *chain.Info
+	client    net.ProtocolClient
+	following bool
+	sync.Mutex
+}
+
+// NewSyncer returns a syncer implementation
+func NewSyncer(l log.Logger, s CallbackStore, info *chain.Info, client net.ProtocolClient) Syncer {
+	return &syncer{
+		store:  s,
+		info:   info,
+		client: client,
+		l:      l,
+	}
+}
+
+func (s *syncer) Syncing() bool {
+	s.Lock()
+	defer s.Unlock()
+	return s.following
+}
+
+func (s *syncer) Follow(c context.Context, upTo uint64, nodes []net.Peer) error {
+	s.Lock()
+	if s.following {
+		s.Unlock()
+		return errors.New("already following chain")
+	}
+	s.following = true
+	s.Unlock()
+	defer func() {
+		s.Lock()
+		s.following = false
+		s.Unlock()
+	}()
+
+	s.l.Debug("syncer", "starting", "up_to", upTo, "nodes", peersToString(nodes))
+
+	last, err := s.store.Last()
+	if err != nil {
+		return err
 	}
-	defer h.l.Debug("sync_reply_leave", addr)
-	if fromRound == 0 {
-		last, err := h.chain.Last()
+	fromRound := last.Round + 1
+	// shuffle through the nodes
+	for _, n := range rand.Perm(len(nodes)) {
+		node := nodes[n]
+		// we quickly pass over different nodes to catchup fast
+		cnode, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```

Comment:
so the connections to each node will time out their contexts in 1 second?

Reply:
Removed that line - it was an experiment I had trouble with some tests - Just add cancel now.